lib/heborn_migration/controller/account.ex

```diff
@@ -1,36 +1,118 @@
+  allows players that typed wrong emails to retry their migrations.
+  """
+  def claim(display_name) do
+    changeset = Claim.create(display_name)
+
+    with \
+      nil <- Repo.get_by(Claim, display_name: display_name),
+      nil <- Repo.get_by(Account, username: String.downcase(display_name)),
+      {:ok, claim} <- Repo.insert(changeset)
+    do
+      {:ok, claim.token}
+    else
+      {:error, changeset} ->
+        {:error, changeset}
+
+      claim = %Claim{} ->
+        {:ok, claim.token}
+
+      account = %Account{} ->
+        account
+        |> Account.expired?()
+        |> maybe_expire_account(account, changeset)
+    end
+  end
+
+  @spec migrate(Claim.t, email :: String.t, password :: String.t) ::
     {:ok, Account.t}
     | {:error, Ecto.Changeset.t}
-  def create(heborning_user, email, password) do
-    heborning_user
-    |> Account.create(email, password)
-    |> Repo.insert()
+  @doc """
+  Migrates account claimed with given `token`, sends an e-mail with the
+  confirmation code.
+  """
+  def migrate(claim, email, password) do
+    Repo.transaction fn ->
+      Repo.delete!(claim)
+
+      result =
+        claim
+        |> Account.create(email, password)
+        |> Repo.insert()
+
+      case result do
+        {:ok, account} ->
+          account
+        {:error, changeset} ->
+          Repo.rollback(changeset)
+      end
+    end
   end
 
-  @spec fetch(pos_integer) ::
+  @spec confirm!(Confirmation.t) ::
     Account.t
-    | nil
-  def fetch(id),
-    do: Repo.get(Account, id)
+  @doc """
+  Confirms an already migrated account.
 
-  @spec fetch_by_username(String.t) ::
-    Account.t
-    | nil
-  def fetch_by_username(username),
-    do: Repo.get_by(Account, username: String.downcase(username))
+  The expiration date doesn't affect this function, as it's just intented to
+  help users that wrongly typed their emails..
+  """
+  def confirm!(confirmation) do
+    {:ok, account} =
+      Repo.transaction fn ->
+        confirmation =
+          confirmation
+          |> Repo.preload(:account)
+          |> Confirmation.confirm()
+          |> Repo.update!()
+
+        Repo.delete!(confirmation)
+
+        confirmation.account
+      end
 
-  @spec confirm(Account.t) ::
-    {:ok, Account.t}
-    | {:error, Ecto.Changeset.t}
-  def confirm(account) do
     account
-    |> Account.confirm()
-    |> Repo.update()
+  end
+
+  # replaces an expired unconfirmed account
+  @spec maybe_expire_account(expire? :: boolean, Account.t, Ecto.Changeset.t) ::
+    {:ok, Claim.token}
+    | {:error, Ecto.Changeset.t}
+  defp maybe_expire_account(true, account, changeset) do
+    Repo.transaction fn ->
+      account = Repo.preload(account, :confirmation)
+
+      Repo.delete!(account.confirmation)
+      Repo.delete!(account)
+
+      case Repo.insert(changeset) do
+        {:ok, claim} ->
+          claim.token
+        {:error, changeset} ->
+          Repo.rollback(changeset)
+      end
+    end
+  end
```

Comment:
Evaluate the use of preload, as it's yet another database query.
Confirmation could be removed with a cascade operation.